yapapi/executor/__init__.py

```diff
@@ -408,34 +399,26 @@ async def start_worker(agreement: rest.market.Agreement) -> None:
             emit(events.WorkerFinished(agr_id=agreement.id))
 
         async def worker_starter() -> None:
-            nonlocal agreements_confirmed
             while True:
                 await asyncio.sleep(2)
-                if (
-                    offer_buffer
-                    and len(workers) < self._conf.max_workers
-                    and work_queue.has_unassigned_items()
-                ):
-                    provider_id, b = random.choice(list(offer_buffer.items()))
-                    del offer_buffer[provider_id]
+                await agreements_pool.cycle()
+                if len(workers) < self._conf.max_workers and work_queue.has_unassigned_items():
                     new_task = None
                     try:
-                        agreement = await b.proposal.create_agreement()
-                        provider = (await agreement.details()).provider_view.extract(NodeInfo)
-                        emit(events.AgreementCreated(agr_id=agreement.id, provider_id=provider))
-                        if not await agreement.confirm():
-                            emit(events.AgreementRejected(agr_id=agreement.id))
+                        new_task = await agreements_pool.use_agreement(
+                            lambda a: loop.create_task(start_worker(a))
+                        )
+                        if new_task is None:
                             continue
-                        emit(events.AgreementConfirmed(agr_id=agreement.id))
-                        agreements_confirmed += 1
-                        new_task = loop.create_task(start_worker(agreement))
                         workers.add(new_task)
                     except CancelledError:
                         raise
-                    except Exception as e:
+                    except Exception:
+                        if self._conf.traceback:
+                            traceback.print_exc()
                         if new_task:
                             new_task.cancel()
-                        emit(events.ProposalFailed(prop_id=b.proposal.id, reason=str(e)))
+                        logger.exception("There was a problem during use_agreement")
```

Comment:
```suggestion
                        logger.debug("There was a problem during use_agreement", exc_info=True)
```
so it has the same logging level as the messages that are produced by emitting events